rust/otap-dataflow/crates/otap/src/encoder.rs

```diff
@@ -5,22 +5,210 @@ use otap_df_pdata_views::views::{
     resource::ResourceView,
+    trace::{
+        EventView, LinkView, ResourceSpansView, ScopeSpansView, SpanView, StatusView, TracesView,
+    },
 };
 use otel_arrow_rust::{
     encode::record::{
         attributes::{AttributesRecordBatchBuilder, AttributesRecordBatchBuilderConstructorHelper},
         logs::LogsRecordBatchBuilder,
+        spans::{EventsBuilder, LinksBuilder, SpansRecordBatchBuilder},
     },
-    otap::{Logs, OtapBatch},
+    otap::{Logs, OtapBatch, Traces},
     otlp::attributes::parent_id::ParentId,
     proto::opentelemetry::arrow::v1::ArrowPayloadType,
 };
 
-use crate::encoder::error::Result;
+use crate::encoder::error::{Error, Result};
 
 mod cbor;
 mod error;
 
+/// Traverse the trace structure within the TracesView and produces an `OtapBatch' for the span
+/// data.
+pub fn encode_spans_otap_batch<T>(traces_view: &T) -> Result<OtapBatch>
+where
+    T: TracesView,
+{
+    let mut resource_attrs = AttributesRecordBatchBuilder::<u16>::new();
+    let mut scope_attrs = AttributesRecordBatchBuilder::<u16>::new();
+    let mut span_attrs = AttributesRecordBatchBuilder::<u16>::new();
+    let mut event_attrs = AttributesRecordBatchBuilder::<u32>::new();
+    let mut link_attrs = AttributesRecordBatchBuilder::<u32>::new();
+
+    let mut curr_resource_id: u16 = 0;
+    let mut curr_scope_id: u16 = 0;
+    let mut curr_span_id: u16 = 0;
+    let mut curr_event_id: u32 = 0;
+    let mut curr_link_id: u32 = 0;
+
+    let mut spans = SpansRecordBatchBuilder::new();
+    let mut events = EventsBuilder::new();
+    let mut links = LinksBuilder::new();
+
+    // First, we traverse the view collecting the trace data into our RecordBatch builders.
+
+    #[allow(clippy::explicit_counter_loop)]
+    for resource_spans in traces_view.resources() {
+        if let Some(resource) = resource_spans.resource() {
+            for kv in resource.attributes() {
+                resource_attrs.append_parent_id(&curr_resource_id);
+                append_attribute_value(&mut resource_attrs, &kv)?;
+            }
+        }
+
+        for scope_spans in resource_spans.scopes() {
+            if let Some(scope) = scope_spans.scope() {
+                for kv in scope.attributes() {
+                    scope_attrs.append_parent_id(&curr_scope_id);
+                    append_attribute_value(&mut scope_attrs, &kv)?;
+                }
+            }
+
+            for span in scope_spans.spans() {
+                // set the resource
+                spans.resource.append_id(Some(curr_resource_id));
+                spans
+                    .resource
+                    .append_schema_url(resource_spans.schema_url().as_deref());
+                spans.resource.append_dropped_attributes_count(
+                    resource_spans
+                        .resource()
+                        .map(|r| r.dropped_attributes_count())
+                        .unwrap_or(0),
+                );
+
```

Comment:
For a backend that deserializes on-the-fly, and to avoid repeating the same processing for each span, it would be better to systematically store the schema_url and dropped_attributes_count defined at the resource level in variables within the enclosing loop. I’m not sure if the same pattern is used for logs, but it might be worth checking.

A similar optimization could be applied for scopes.

Reply:
That's a great suggestion! Done!.